### PR TITLE
group test

### PR DESCRIPTION
**Pull request by diegolucasilva**
pending:: redis embedded

### CODE REVIEW

**Comment by padovese** [comment]
mongo*

File: src/test/kotlin/com/maitre/menuservice/adapter/group/in/get/GetGroupHandlerTest.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+internal class GetGroupHandlerTest{
+
+    @Autowired
+    lateinit var webTestClient: WebTestClient
+
+    @Test
+    fun `Should get a groups by menu_id`() {
+        webTestClient
+                .get()
+                .uri("/group?menu_id=MENU_af60830b-d190-43bf-afcb-f5cc2656ea25")
```
Comment (padovese) [inline src/test/kotlin/com/maitre/menuservice/adapter/group/in/get/GetGroupHandlerTest.kt]: mocka esse get, usa o do menu como exemplo, ta no PR #25

**Review by padovese — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/maitre/menuservice/adapter/group/in/update/UpdateGroupHandlerImpl.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                    )
+                }
+                .flatMap {
+                    updateGroupUseCase.execute(it.toDomain(), serverRequest.pathVariable("id"))
+                }
+                .map {
+                    it.toResponseDTO()
+                }.flatMap {
+                    logger.info("ResponseBody=${it.toJson()}")
+                    created(URI("IMPLEMENTAR_HATEOS_SELF${serverRequest.path()}/${it.id}"))
```
Comment (padovese) [inline src/main/kotlin/com/maitre/menuservice/adapter/group/in/update/UpdateGroupHandlerImpl.kt]: The resource is already created. A better response for this use case should be a 200 OK.

File: src/main/kotlin/com/maitre/menuservice/domain/group/entity/Group.kt [hunk 3]
```diff
@@ -1,14 +1,14 @@
 package com.maitre.menuservice.domain.group.entity
 
-import java.util.UUID
+import java.util.*
```
Comment (padovese) [inline src/main/kotlin/com/maitre/menuservice/domain/group/entity/Group.kt]: Don't use wildcards.

File: src/main/kotlin/com/maitre/menuservice/exception/GroupNotFoundException.kt [hunk 4]
```diff
@@ -1,3 +1,3 @@
 package com.maitre.menuservice.exception
 
-class GroupNotFoundException(id: String) : RuntimeException("Group not found. menu_id=$id")
\ No newline at end of file
+class GroupNotFoundException(id: String) : RuntimeException("Group not found. $id")
```
Comment (padovese) [inline src/main/kotlin/com/maitre/menuservice/exception/GroupNotFoundException.kt]: Let `group_id=$id`
The logs engines like splunk, kibana and cloudWatch use key=value format to index their logs.

File: src/test/kotlin/com/maitre/menuservice/adapter/group/in/create/CreateGroupHandlerTest.kt [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+import com.maitre.menuservice.adapter.group.`in`.dto.GroupRequestDTO
+import com.maitre.menuservice.domain.group.entity.GroupType
+import org.junit.jupiter.api.Test
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.http.MediaType
+import org.springframework.test.web.reactive.server.WebTestClient
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+ class CreateGroupHandlerTest {
```
Comment (padovese) [inline src/test/kotlin/com/maitre/menuservice/adapter/group/in/create/CreateGroupHandlerTest.kt:19]: Use test profile only for integration testes. And rename/realocate it to indicate that is an integration test.

Comment (padovese) [reply src/test/kotlin/com/maitre/menuservice/adapter/group/in/create/CreateGroupHandlerTest.kt:19]: Also, mock the repository using whenever().then()

File: src/test/kotlin/com/maitre/menuservice/adapter/group/in/get/GetGroupHandlerTest.kt [hunk 6]
```diff
@@ -0,0 +1,45 @@
+package com.maitre.menuservice.adapter.group.`in`.get
+
+import com.maitre.menuservice.domain.group.entity.GroupType
+import org.junit.jupiter.api.Test
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.test.web.reactive.server.WebTestClient
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+internal class GetGroupHandlerTest{
```
Comment (padovese) [inline src/test/kotlin/com/maitre/menuservice/adapter/group/in/get/GetGroupHandlerTest.kt]: Same as above.

File: src/test/kotlin/com/maitre/menuservice/domain/group/usecases/CreateGroupUseCaseTest.kt [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        givenSomeCorrectSavePort()
+
+        StepVerifier.create(whenUseCaseIsExecuted())
+                .expectNextCount(1)
+                .verifyComplete()
+    }
+
+    private fun givenSomeCorrectSavePort() {
+        whenever(saveGroupPort.save(any<Group>())).thenReturn(Mono.just(group))
+        whenever(saveGroupPort.save(any<Group>())).thenReturn(Mono.just(group))
```
Comment (padovese) [inline src/test/kotlin/com/maitre/menuservice/domain/group/usecases/CreateGroupUseCaseTest.kt]: duplicated lines.

File: src/test/kotlin/com/maitre/menuservice/domain/group/usecases/CreateGroupUseCaseTest.kt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    private val getMenuPort: GetMenuPort = mock()
+    private val createGroupUseCase = CreateGroupUseCase(logger, saveGroupPort, getMenuPort)
+    @Test
+    fun `Test Create Group Use Case - verify if save port is being called`() {
+        givenSomeCorrectMenuPort()
+        givenSomeCorrectSavePort()
+
+        whenUseCaseIsExecuted()
+
+        verify(saveGroupPort, times(1)).save(any<Group>())
```
Comment (padovese) [inline src/test/kotlin/com/maitre/menuservice/domain/group/usecases/CreateGroupUseCaseTest.kt:28]: you can also verify if the getMenuPort is being called.

File: src/test/kotlin/com/maitre/menuservice/domain/group/usecases/GetGroupsByMenuIdUseCaseTest.kt [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        verify(getGroupsByMenuIdPort, times(1)).getByMenuId(any<String>())
+    }
+
+    @Test
+    fun `Test Get Group Use Case - verify if an event is emitted`() {
+        givenSomeCorrectGetPort()
+
+        StepVerifier.create(whenUseCaseIsExecuted())
+                .expectNextCount(1)
+                .verifyComplete()
```
Comment (padovese) [inline src/test/kotlin/com/maitre/menuservice/domain/group/usecases/GetGroupsByMenuIdUseCaseTest.kt]: This test is ok, but as a tip, if you are working with flux, you can mock a list of N and then verify a count of N events emitted.

**Review by padovese — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/maitre/menuservice/domain/group/entity/Group.kt (hunk 3 above)
Comment (diegolucasilva) [reply src/main/kotlin/com/maitre/menuservice/domain/group/entity/Group.kt]: done

File: src/main/kotlin/com/maitre/menuservice/exception/GroupNotFoundException.kt (hunk 4 above)
Comment (diegolucasilva) [reply src/main/kotlin/com/maitre/menuservice/exception/GroupNotFoundException.kt]: This exception is used in 2 situations. I think we can refactor all this type of exception and create only ResourceNotFoundException that can receive a complete message.

File: src/test/kotlin/com/maitre/menuservice/adapter/group/in/create/CreateGroupHandlerTest.kt (hunk 5 above)
Comment (diegolucasilva) [reply src/test/kotlin/com/maitre/menuservice/adapter/group/in/create/CreateGroupHandlerTest.kt:19]: mock doesnt work... pending a issue later

File: src/test/kotlin/com/maitre/menuservice/adapter/group/in/get/GetGroupHandlerTest.kt (hunk 6 above)
Comment (diegolucasilva) [reply src/test/kotlin/com/maitre/menuservice/adapter/group/in/get/GetGroupHandlerTest.kt]: done

File: src/test/kotlin/com/maitre/menuservice/domain/group/usecases/CreateGroupUseCaseTest.kt (hunk 7 above)
Comment (diegolucasilva) [reply src/test/kotlin/com/maitre/menuservice/domain/group/usecases/CreateGroupUseCaseTest.kt]: done

File: src/test/kotlin/com/maitre/menuservice/domain/group/usecases/CreateGroupUseCaseTest.kt (hunk 8 above)
Comment (diegolucasilva) [reply src/test/kotlin/com/maitre/menuservice/domain/group/usecases/CreateGroupUseCaseTest.kt:28]: done

File: src/main/kotlin/com/maitre/menuservice/adapter/group/in/update/UpdateGroupHandlerImpl.kt (hunk 2 above)
Comment (diegolucasilva) [reply src/main/kotlin/com/maitre/menuservice/adapter/group/in/update/UpdateGroupHandlerImpl.kt]: done

**Review by padovese — APPROVED** [review]

**Pull request merged**